Add GEP-1364 proposal (#1383)
kubernetes-sigs/gateway-api: site-src/geps/gep-1364.md

# GEP-1364: Status and Conditions Update

* Issue: [#1364](https://github.com/kubernetes-sigs/gateway-api/issues/1364)
* Status: Implementable

(See definitions in [Kubernetes KEP][kep-status].

## TLDR

> The status, particularly the Conditions, across the whole Gateway API has very much grown
> grown organically, and so has many inconsistencies and odd behaviors.

Reviewer: ```suggestion
The status, particularly the Conditions, across the whole Gateway API has very much grown
organically, and so has many inconsistencies and odd behaviors.
```

Reviewer 2: Tiny nit: has -> have

This GEP covers doing a review and consolidation to make Condition behavior consistent
across the whole API.

## Goals

* Update Conditions design to be consistent across Gateway API resources
* Provide a model and guidelines for Conditions for future new resources
* Specify changes to conformance required for Condition updates

## Non-Goals

> 

Reviewer 2: ```suggestion
* Define the full set of Conditions that will ever be used with Gateway API
```

## Introduction

Gateway API currently has a lot of issues related to status, especially that
status is inconsistent ([#1111][1111]), that names are hard to understand ([#1110][1110]),

> Reasons aren't explained properly ([#1362][1362]), and so on.

Reviewer 2: Tiny nit (feel free to ignore):
```suggestion
and that Reasons aren't explained properly ([#1362][1362]).
```

As the API has grown, the way we talk about resources has changed a lot, and some of the
status design hasn't been updated since resources were created.

So, for example,  we have GatewayClass with `Accepted`, Gateway with `Scheduled`,
the Gateway Listeners with `Detached` (which you want to be `false`, unlike the previous
two), and then Gateways and Gateway Listeners have `Ready`, but Route doesn't (and which
also you want to be `true`).

This document lays out large-scale changes to the way that we talk about resources,
and the Conditions to match them. This means that there will be an unavoidable break
in what constitutes a healthy or unhealthy resource, and there will be changes

> required for all conformant implementations to stay conformant.

Reviewer 2: Maybe specifically mention the concept of being conformant with a specific release of the API?
```suggestion
required for all implementations to be conformant with the release that includes these changes.
```

This level of change is not optimal, and the intent is to make this a one-off change
that can be built upon for future resources - since there are definitely more resources
on the way.

## Background: Kubernetes API conventions and prior art on Conditions

Because this GEP is mainly concerned with updating the Conditions we are setting in
Gateway API resources' `status`, it's worth reviewing some important points about

> Conditions. (Thi information is mainly taken from the [Typical status properties][typstatus]

Reviewer 3: Missing href for the link here?

Author: It's all the way down at the bottom of the doc, with the other references. I thought it was better to have everything in the one place.

> Conditions. (Thi information is mainly taken from the [Typical status properties][typstatus]

Reviewer 2: ```suggestion
Conditions. (This information is mainly taken from the [Typical status properties][typstatus]
```

section of the API conventions document.)

* Conditions are a standard type used to represent arbitrary higher-level status from
a controller.
* They are a listMapType, a list that is enforced by the apiserver to have only
one entry of each item, using the `type` field as a key. (So, this is effectively
a map that looks like a list in YAML form).
* Each has a number of fields, the most important of which for this discussion

> are `type`, `status`, `season`, and `observedGeneration`.

Reviewer 3: ```suggestion
are `type`, `status`, `reason`, and `observedGeneration`.
```

  * `type` is a string value indicating the Condition type. `Accepted`, `Scheduled`,
  and `Ready` are current examples.
  * `status` indicates the state of the condition, and can be one of three values,
  `true`, `false`, or `unknown`. Unkown in particular is important, because it
  means that the controller is unable to determine the status for some reason.
  * `reason` is a CamelCase string that is a brief description of the reason why
  the `status` is set the way it is.
  * `observedGeneration` is an optional field that sets what the `metadata.generation`
  field was when the controller last saw a resource.
* Conditions shoud describe the _current state_ of the resource at observation
time, which means that they should be an adjective (like `Ready`), or a past-tense

> verb (like `Attached`).

Reviewer 2: Would `Attaching` ever be valid as a condition? Maybe just verb or adjective is sufficient here?

Author: That sort of usage is specifically called out as not desirable in the API conventions, and I agree. Conditions are supposed to be level-triggered, not edge-triggered, so having a condition that says that something is happening is a bit weird. (Should `Attaching` transition to `false` once the attaching process is complete?)

Reviewer: For posterity, can we cite some sources for that assertion? It would probably actually be good as a part of this status effort to document somewhere _in the code_ that we're avoiding this due to conventions, and then point to those conventions, as this will help future contributors to more quickly avoid pursuing such things (or give them a path to who they need to convince in the greater kubernetes community).

Author: This is almost a direct quote from the API conventions as it is, should I put a link in here to the exact line?

Author: The exact quote is as follows, I can just inline if it's better, but I was trying to avoid just pasting in the entire section:

---

Condition type names should describe the current observed state of the resource, rather than describing the current state transitions. This typically means that the name should be an adjective ("Ready", "OutOfDisk") or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb ("Deploying"). Intermediate states may be indicated by setting the status of the condition to Unknown.

For state transitions which take a long period of time (e.g. more than 1 minute), it is reasonable to treat the transition itself as an observed state. In these cases, the Condition (such as "Resizing") itself should not be transient, and should instead be signalled using the True/False/Unknown pattern. This allows other observers to determine the last update from the controller, whether successful or failed. In cases where the state transition is unable to complete and continued reconciliation is not feasible, the Reason and Message should be used to indicate that the transition failed.

---

I don't believe that we generally have any process that we want to handle in the way called out in the exception bullet point. Happy to be guided here on what you'd like to see.

negative-polarity conditions.

There is also some extra context that's not in the API conventions doc:

SIG-API Machinery has been reluctant to add fields that would aid in machine-parsing
of Conditions, especially fields that would indicate the polarity, because they
are intended more for human consumption than machine consumption.

This means that there's no guidance from upstream about condition polarity. We'll
discuss this more when we talk about new conditions.

The guidance about Conditions being added as soon as a controller sees a resource
is a bit unclear - as written in the conventions, it seems to imply that _all_ 
relevant conditions should always be added, even if their status has to be set to
`unknown`.
Gateway API resources do not currently require this, and the practice seems to be
uncommon.

## Proposed changes

### Proposed changes summary

* All the current Conditions that indicate that the resource is okay and ready
for processing witll be replaced with `Attached`, except for GatewayClass (since
it is the root of the resource tree, it will stay with `Accepted`).
* In general, resources should be considered `Attached` if their config will

> generate some config in the underlying data plane.

Reviewer 3: It might be helpful to be explicit here:
```suggestion
generate some config in the underlying data plane (whether that config is valid or not).
```

Author: I've been assuming that only valid config will be passed to the data plane. If a resource would generate invalid config, then to my mind it shouldn't be considered attached.

Again, here I'm thinking of HTTPRoute adding 500s for things that would otherwise cause a resource to become unattached (that is, it doesn't  match anything, or it doesn't have a valid `backendRef`, or there are zero endpoints in a valid backendRef). For Route resources that don't have an error code response like HTTP does, presence of that class of error should probably result in it becoming unattached (since you can't generate any meaningful data plane config at all).

* All Conditions will be positive polarity (`GoodState: true`, not `BadState: false`)

> * All relevant Conditions for a resource must be added when it's observed.

Author: So, does this mean that all conditions that have Core conformance should always be present? Or does this mean that a set of positive-polarity summary endpoints we name should always be present? The former could include any negative-polarity conditions we define as having Core conformance (which would clash with the "don't have double negatives" guidance we're looking for in other places.)

> * All relevant Conditions for a resource must be added when it's observed.

Author: I've guessed "summary conditions" here - meaning just the positive polarity ones.

So, when we talk about Gateway API, it's _always_ about a _system of related resources_.

We already acknowledge this when we talk about Routes "attaching" to Gateways, or Gateways
referencing Services, or Gateways requiring a GatewayClass in their spec.

So, this GEP proposes that we formalize this by leaning into the word "attach" as the
primary verb to describe the relationships between resources in the system.

So, Gateways _attach_ to GatewayClasses, Routes _attach_ to Gateways, and in a
bit of an extension, ReferenceGrants _attach_ to the things they are Granting
Reference access to.

Of course, because we're using all of this configuration to describe some sort of data
path from "outside"/lacking cluster context to "inside"/enriched with cluster context,
we also need a way to describe when that data path is configured and working.

Again, we already have a word we use for this - "Ready", so this GEP proposes keeping
that and tightening up the definition.

### Condition polarity

In terms of the polarity of conditions, we have three options, of which only two are
really viable:
* All conditions must be negative polarity
* All conditions must be positive polarity

> * Some conditions can be positive polarity, but most should be negative.

Reviewer 2: My own personal preference would be for conditions with a positive polarity to represent the normal healthy state and always be present, and for conditions with a negative polarity to represent very specific error states and to only be present when true. I think we should do everything we can to avoid the presence of double negatives like "Detached == False", but recognize that there are some error states that are best represented with a negative condition. I just don't think those negative conditions should be present when false.

Author: If we're going to have a mixed-polarity set of conditions, then I think we need to define:
- which Conditions are positive polarity (Probably `Attached`, `Programmed`, `Ready`, and `Valid` if we do it)
- which Conditions are positive polarity and should always be present (Probably `Attached`, `Programmed`, and `Valid` if we do it).
- which Conditions are core (currently, everything other than `Ready`)
- which Conditions are extended (only `Ready`).
- which Conditions are negative (all the error conditions).

The fact that the user experience of `Ready` or conditions like `Accepted` being `true`
in the healthy case is much better rules out the first option, so we are left to
decide between enforcing that all conditions are positive, or that we have a mix.

> Having an arbitrary mix will make doing machine-based extraction of information

Author: It means that anything consuming the Conditions will need to have a hard-coded list of the positive polarity conditions. Not a deal breaker, but it's another obstacle in the way of building a generic condition-parsing library.

plane.

So, All Conditions Positive pros:
* We're close already. Most conditions in the API are currently positive polarity.
* Easier to understand - there are no double negatives. "Good: true" is less
cognitive overhead than "NotGood: false".

Cons:
* Reduces flexibility - it can surprisingly difficult to avoid double negatives for
conditions that describe error states, as in general programmers are more used
to reporting "something went wrong" than they are "everything's okay".

Not sure if pro or con:
* Leans the design towards favoring conditions always being present, since you
can't be sure if everything is good unless you see `AllGood: true`. The absence
of a positive-polarity condition implies that the condition could be false. This
puts this option more in line with the API guidelines on this point.

#### Some Conditions Positive

In this case, only a limited set of summary conditions are positive, and the rest
are negative.

Pros:
* Error states can be described with `Error: true` instead of `NoError: false`.
* Negative polarity error conditions are more friendly to not being present (since
absence of `Error: true` implies everything's okay).

Cons:

> * Double negatives in the happy case: `NoError: false`

Reviewer 2: This seems like a con of "All Conditions Positive" but not sure it applies when we allow both positive and negative conditions?

Author: Ah yes, you're right.

#### Decision

Gateway API resource conditions will all be positive polarity.

### Should conditions always be added?

Since we are mandating that conditions should always be positive, we should also
specify that controllers always place all conditions on each resource when they
reconcile it, even if that update sets the status to `unknown` because of

> information not being available.

Reviewer 2: I think this approach would lead to a _very_ noisy conditions list. I'd personally prefer leaving room for negative/error conditions to be unset unless they were true.

Author: There's only one struct allowed for Conditions, so we can't change nil-values. What we'll need to say is "For all conditions other than <this list>, they must be negative polarity, and must only be present if they are status: true". Conversely, this also means that the absent value for these Condition types is "False", not "I don't know yet". Not the end of the world, but a little different to the API conventions.

condition means that a user can't be _sure_ that the condition is okay. So, we
should ensure that all the positive conditions are always present, even if their
status is `unknown`.

### New and Updated Conditions

#### `Attached`

This GEP proposes replacing all conditions that indicate syntactic and semantic
validity with one, `Attached` condition type, with the exception of the
GatewayClass resource (because it's the root of the resource tree and doesn't
attach to anything).

That is, the proposal is to replace:
* `Scheduled` on Gateway
* `Detached` on Listener
* `Accpeted` on Route

with `Attached` in all these locations.

GatewayClass will maintain the `Accepted` condition.

All of these conditions share the following meanings:
* The resource has been accepted for processing by the controller
* The resource is syntatically and semantically valid, and internally consistent
* The resource fits into a larger system of Gateway API resources, and there is
is no missing information, including but not limited to:
  * Any mandatory references resolve to existing resources (examples here are the
  Gateway's gatewayClass field, or the `parentRefs` field in Route resources)

>   * Any specified TLS secrets exist

Reviewer 4: Currently the API supports partially-valid resources. I believe this is proposing a partially-valid resource would be `Attached=false`, but the config is actually 'active' (or, I would describe it, "attached").

To me this feels a bit odd. The old API sort of had this issue as well.

Thinking purely from a pedantic-correctness POV (which may not be ideal - sometimes simpler is better), I would expect there to be two conditions (I use bad names intentionally to avoid confusion): "ThisResourceWasApplied" - any part of the rule had some effect on the system, and "ThisRuleIsEntirelyValid" - no errors found in the rule at all.

Its possible my concern would be addressed with a different word than "Attached", rather than more conditions, as well

WDYT?

Reviewer 4: I guess this is somewhat addressed below.. but I am not sure its accurate? It seems we are defining a very small set of "failures that do not become unattached", but there are a lot more partially-accepted resources I think?

Author: The guideline I've been basing the "does something become unattached?" on is "Does this produce at least some configuration in the underlying data plane?". So I guess it sounds like I need to be more explicit with that guideline, but that aside from that, we're largely in agreement?

Reviewer 4: I think it makes sense to have "Is this attached" and "is this 100% valid" as two separate things. 

Part of my confusion, I think, is I am not actually sure which fields result in "not attached". It seems _most_ issues are partially accepted. Which is fine - `Attached` just becomes basically a pretty high level acknowledgement that the controller is handling the resource

Reviewer 3: Yeah I think I agree with @howardjohn's point here: "does this resource produce config" is distinct from "does this resource produce _valid_ config". I'm sure some implementations don't allow invalid (e.g. duplicate route names or something) to be sent to the data plane, but it doesn't feel like that's a requirement for the Gateway API at this juncture.

Author: I guess I wrote this pooly - by "Does this resource produce config" I meant to read "Is this resource valid enough to produce some config when combined with the rest of the resource set".

The thing I'm thinking of particularly here is that for HTTPRoute, invalid references still produce data path config because the match will respond with 500s. Whereas for TCPRoute etc, if you don't have a service to route to, there's no way you can build a forwarding path - there's nothing to forward to! So that won't produce any data path config at all, and so, in the current design, would cause the resource to fail to be attached.

I think that partial validity is an enormous minefield of weirdness, just like not applying config if it would break things.

Reviewer 5: Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

Reviewer 6: > Is it possible for something like an HTTPRoute to define two paths, one of which is "valid enough" config, while the other is garbage?

Most cases of an invalid configuration in one route rule (with others being completely valid within the same HTTPRoute) seem to be misconfigurable only in the context of a reference to another resource and fall into the `ResolvedRefs` bucket (which seems stated below and possibly agreed upon that `Attached: true` be set with `ResolvedRefs: false` and return HTTP code 500 for the invalid rule).

One such example not covered is if you were to use a `HTTPRequestRedirectFilter` and `HTTPURLRewrite` in the list of filters for an individual `HTTPRouteRule` ([ref](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRequestRedirectFilter)). At the moment it doesn't appear we have prescribed any status code or other error handling for this error. This seems to fall into the bucket of things that cannot produce data plane configuration.

Going by the statements below, if this rule were the only one present `Attached: false` should be set, since it cannot result in any configuration (with another more specific error condition). If present with other valid rules, `Attached: true` should be set (with another more specific error condition again). I can definitely see the argument for setting `Attached: false` however in the latter case, since stated above, the `Attached` condition means `The resource is syntatically and semantically valid, and internally consistent`.

It seems that the `Valid: true` condition mentioned above ends up being somewhat equivalent to having  for example for HTTPRoute `ResolvedRefs: true` and no error conditions (and vice versa, `Valid: false` equivalent to `ResolvedRefs: false` *or* some more specific error condition). Maybe this framing helps a little with determining possible merits of having another condition?

> If so, are there expectations or requirements from a conformance PoV ("always apply all or none") that we need to specify?

Just doing a quick pass, I don't *think* I saw any existing conformance tests that cover this kind of thing so I agree that would be good to add.

Author: Avoiding "always apply all or none" is the reason this is tricky. There are mixed expectations about what happens if _some_ config is present but bad, and I'm attempting to thread the needle here and make this as user-friendly as possible while not becoming really difficult to implement or understand.

Author: I added a section on partial validity to hopefully make this clearer.

* The resource is supported by the controller by ensuring things like:
  * Any Kinds being referred to by the resource are supported
  * Features being used by the resource are supported

All of these rules can be summarized into:
* The resource will produce some configuration in the underlying data plane.

For Gateway, `Attached` also subsumes the functions of `Scheduled`: `Attached`
set to `true` means that sufficient capacity exists on underlying infrastructure
for the Gateway to be provisioned. If that capacity does not exist, then the
Gateway cannot be reconciled successfully, and so fails to attach to the
owning GatewayClass.

Note that some classes of inter-resource reference failure do _not_ cause a resource
to become unattached (that is, to have the `Attached` condition set to `status: false`).

> * Nonexistent Service backends - if the backend does not exist on a HTTPRoute that

Reviewer 3: This starts painting an observability problem IMO.  Even if conditions are only meant to be human-readable, how does a human look at an HTTPRoute resource and know "oops I have a typo in my service backend"? Yes, 500s indicate something's wrong, but I would expect to find a condition pointing to the specific issue

Reviewer 7: We should allow room for error conditions in addition to attached in this case so the intformation can be conveyed.

Author: Yes, I guess I should make this clearer, that in this case, it's absolutely expected that there be some additional, negative-polarity error conditions that tell you that the backendRef doesn't exist. (This is a distinct case from the backend existing, but having zero endpoints, which would _not_ generate a "backendRef doesn't exist" error, but would end up with the same behavior).

Reviewer 5: You might also get a `BackendRefEmpty` condition...

for indicating a server failure (like the HTTP code 500 does), not having existing
backends may not produce any configuration in the data plane, and so may cause
the resource to fail to attach. (An example here could be a TCP Route with
no backends, we need to decide if that means that a port should be opened that
actively closes connections, or if no port should be opened.)

#### Ready

Currently, the `Ready` condition text for Gateway says:
```go
	// This condition is true when the Gateway is expected to be able
	// to serve traffic. Note that this does not indicate that the
	// Gateway configuration is current or even complete (e.g. the
	// controller may still not have reconciled the latest version,
	// or some parts of the configuration could be missing).
```

This is pretty unclear - how can the Gateway serve traffic if config is missing?
In the past, we've been asked to have a Condition that only flips to `true` when
*all* required configuration is present.

For many implementations (certainly for Envoy-based ones), getting this information
correctly and avoiding races on applying it is surprisingly difficult. For this

> reason, this GEP proposes we give the `Ready` condition a short window

Reviewer 2: I think I'd rather reserve `Ready` for "Ready at this moment" and instead use a different term to represent what we're actually describing - this config has been written to the data plane but we're not sure how long that will take to propagate. Previously `Reconciled` has been proposed to communicate this.

Author: Okay, I'll update this section to:
- add `Ready` as an Extended support condition, that, if present, indicates that the data plane is fully configured. But because it's not always present, we'll need `Reconciled` to indicate what we have here - that everything is okay and the data plane will be ready soon.

This condition needs a few changes:

* Each resource should have a `Ready` condition
* The `Ready` condition should be clear as to *when* the traffic is able to pass
  the implementation. Since it's _extremely_ difficult to _guarantee_ this, we're
  stuck with some sort of "will be ready in no longer than x seconds" standard
  (which we'll also need to write conformance tests for, probably including a 
  stress test of adding a *lot* of records as well.) As an initial ballpark,

>   I propose 5 seconds.

Reviewer 4: This probably won't work for cloud providers (at least GKE is slower, not sure about others).

I am also not sure it provides value. What do we lose if we don't define anything here?

If we consider something like Knative, they certainly aren't going to be able to just wait X seconds; I'd expect the same with other users

Reviewer 4: And if we can't provide any guarantees about Readiness, then it seems like Ready becomes no different than Attached...

Author: We certainly can mandate `Ready` to be an _extended_ condition, and say "If you supply a Ready condition, it has to work like this". Maybe doing that is a better place to start with `Ready`?

Reviewer 4: Do we have a feel for which implementations can and cannot support `Ready` properly?

Reviewer 2: Agree that cloud providers will likely have trouble implementing a `Ready` condition. Could we introduce a `Reconciled` condition that would be more broadly implementable?

Reviewer 7: +1 any distributed (e.g. geo distributed) implementation has a hard time with this (think: config was pushed to 95% of the sites around the world, is it ready?)

across the whole API.

## Goals

* Update Conditions design to be consistent across Gateway API resources
* Provide a model and guidelines for Conditions for future new resources
* Specify changes to conformance required for Condition updates

## Non-Goals

* Define the full set of Conditions that will ever be used with Gateway API

## Introduction

Gateway API currently has a lot of issues related to status, especially that
status is inconsistent ([#1111][1111]), that names are hard to understand ([#1110][1110]),
and that Reasons aren't explained properly ([#1362][1362]).

As the API has grown, the way we talk about resources has changed a lot, and some of the
status design hasn't been updated since resources were created.

So, for example, we have GatewayClass with `Accepted`, Gateway with `Scheduled`,
the Gateway Listeners with `Detached` (which you want to be `false`, unlike the previous
two), and then Gateways and Gateway Listeners have `Ready`, but Route doesn't (and which
also you want to be `true`).

This document lays out large-scale changes to the way that we talk about resources,
and the Conditions to match them. This means that there will be an unavoidable break

> in what constitutes a healthy or unhealthy resource, and there will be changes

Reviewer 8: Given the beta status of the API - I would keep the old behavior in place and not break anything, maybe until v1.
Implementations generating both old and new is a pain - but far less than impacting users.

Author: I added a section about this just below.

verb (like `Attached`).
5. Conditions should be applied to a resource the first time the controller sees
the resource. This seems to imply that _all conditions should be present on every
resource owned by a controller_, but the rest of the conventions don't make this
clear, and it is often not complied with.
6. It's helpful to have a top-level condition which summarizes more detailed conditions.
The guidelines suggest using either `Ready` for long-running processes, or `Succeeded`
for bounded exection.

From these guidelines, we can see that Conditions can be either _positive polarity_
(healthy resources have them as `status: true`) or _negative polarity_ (healthy
resources have them as `status: false`). `Ready` is an example of a positive polarity
condition, and conditions like `Conflicted` from Listener or `NetworkUnavailable`,
`MemoryPressure`, or `DiskPressure` from the Node resource are examples of
negative-polarity conditions.

There is also some extra context that's not in the API conventions doc:

SIG-API Machinery has been reluctant to add fields that would aid in machine-parsing
of Conditions, especially fields that would indicate the polarity, because they

> are intended more for human consumption than machine consumption.

Author: I think it's more that API Machinery has said specifically that they won't add a field for indicating the polarity of a Condition to the Condition struct. This was in a discussion raised by @evankanderson, but I can't seem to find the reference now.

This means that there's no guidance from upstream about condition polarity. We'll
discuss this more when we talk about new conditions.

The guidance about Conditions being added as soon as a controller sees a resource
is a bit unclear - as written in the conventions, it seems to imply that _all_ 

> relevant conditions should always be added, even if their status has to be set to

Reviewer 8: I think it is reasonable for the controller to do this - as an indication the resource is in use by the controller.

It becomes problematic if a resource may be processed by multiple controllers - for example the discussion
of HttpRoute applying to both mesh and Gateway can only work if the type has a prefix (MeshReady). 

In general it would be good to indicate how to handle resources where multiple controllers operate on. While 
it's good to know which controller mess with the resource - maybe the controller should skip this step if the
resource is not ignored ( a mesh controller will ignore any route that don't target both mesh and gateway as example)

Author: In most cases, the Conditions slice is in a separate struct that's part of a slice that's namespaced by controller name - so that controllers should be able to update only their own Conditions. I'd expect that mesh controllers should have a distinct controller name string, so this method should continue working.

Reviewer 2: I think our practice of including `controllerName` for each distinct set of conditions should largely allow multiple controllers to cooperate on status for the same resource, but open to other potential improvements.

Reviewer 5: How do `controllerName` conditions roll up to `Accepted`?

Also, controllers probably need to be sure to use PATCH on the `status` sub-resource or do conditional PUTs and avoid extra writes if the conditions haven't changed.

Author: For some resources, all conditions are namespaced by `controllerName`. I agree we need to ensure that the rollup behaivor is specified. I'm going to add a section about partial validity that should help.

`unknown`.
Gateway API resources do not currently require this, and the practice seems to be
uncommon.

## Proposed changes

### Proposed changes summary

* All the current Conditions that indicate that the resource is okay and ready

> for processing witll be replaced with `Attached`, except for GatewayClass (since

Reviewer 8: Sound good - but I think Gateway may need additional info, like Ready - since the process may involve 2 steps.
For example Attached will be set when the controller generates the configs - but Ready when certificates are provisioned. CertManager would watch for Attached and start the ACME process, create the Secret - and the gateway controller would switch to Ready.

Author: I'm proposing a `Programmed` condition to summarize this without the "this is ready right now" implications of Ready. I'll put those changes into this summary.

* Negative polarity error conditions are more friendly to not being present (since
absence of `Error: true` implies everything's okay).

Cons:
* Any code handling conditions will need a list of the positive ones, and will
need to assume that any others are negative.

#### Decision

Gateway API conditions will be positive for conditions that describe the happy
state of the object, which is currently `Attached` and `ResolvedRefs`.

### Should conditions always be added?

Not all of them.

Positive polarity Conditions that describe the desirable state of the object must
always be set. These are currently `Attached`, `ResolvedRefs`, and whatever we
end up choosing to represent "the data path will be done soon", `Ready`, `Programmed`,
or other.

### New and Updated Conditions

#### `Attached`

This GEP proposes replacing all conditions that indicate syntactic and semantic
validity with one, `Attached` condition type, with the exception of the
GatewayClass resource (because it's the root of the resource tree and doesn't
attach to anything).

That is, the proposal is to replace:
* `Scheduled` on Gateway
* `Detached` on Listener

> * `Accpeted` on Route

Reviewer 9: ```suggestion
* `Accepted` on Route
```

Author: Thanks!

Conditions. (This information is mainly taken from the [Typical status properties][typstatus]
section of the API conventions document.)

1. Conditions are a standard type used to represent arbitrary higher-level status from
a controller.
2. They are a listMapType, a list that is enforced by the apiserver to have only
one entry of each item, using the `type` field as a key. (So, this is effectively
a map that looks like a list in YAML form).
3. Each has a number of fields, the most important of which for this discussion
are `type`, `status`, `reason`, and `observedGeneration`.

    * `type` is a string value indicating the Condition type. `Accepted`, `Scheduled`,
    and `Ready` are current examples.
    * `status` indicates the state of the condition, and can be one of three values,
    `true`, `false`, or `unknown`. Unkown in particular is important, because it
    means that the controller is unable to determine the status for some reason.
    * `reason` is a CamelCase string that is a brief description of the reason why
    the `status` is set the way it is.
    * `observedGeneration` is an optional field that sets what the `metadata.generation`

>     field was when the controller last saw a resource.

Reviewer 2: I think we'd had some conversations about making this required for conformant Gateway implementations. Probably out of scope for this GEP, just wanted to make sure it was intentionally excluded.

Author: Good point, I updated.

4. Conditions shoud describe the _current state_ of the resource at observation
time, which means that they should be an adjective (like `Ready`), or a past-tense
verb (like `Attached`). This one in particular is documented pretty closely in the
[Typical status properties][typstatus] section of the guidelines.
5. Conditions should be applied to a resource the first time the controller sees
the resource. This seems to imply that _all conditions should be present on every

> resource owned by a controller_, but the rest of the conventions don't make this

Reviewer 2: I agree that we should have a set of permanent and reliable conditions that are always present, but I personally don't think it makes sense for error conditions to always be present, especially since that tends to lead to confusing double negative states.

resources have them as `status: false`). `Ready` is an example of a positive polarity
condition, and conditions like `Conflicted` from Listener or `NetworkUnavailable`,
`MemoryPressure`, or `DiskPressure` from the Node resource are examples of
negative-polarity conditions.

There is also some extra context that's not in the API conventions doc:

SIG-API Machinery has been reluctant to add fields that would aid in machine-parsing
of Conditions, especially fields that would indicate the polarity, because they
are intended more for human consumption than machine consumption.

This means that there's no guidance from upstream about condition polarity. We'll
discuss this more when we talk about new conditions.

The guidance about Conditions being added as soon as a controller sees a resource
is a bit unclear - as written in the conventions, it seems to imply that _all_ 
relevant conditions should always be added, even if their status has to be set to
`unknown`.
Gateway API resources do not currently require this, and the practice seems to be
uncommon.

## Proposed changes

### Proposed changes summary

* All the current Conditions that indicate that the resource is okay and ready

> for processing witll be replaced with `Attached`, except for GatewayClass (since

Reviewer 2: ```suggestion
for processing will be replaced with `Attached`, except for GatewayClass (since
```

> it is the root of the resource tree, it will stay with `Accepted`).

Reviewer 10: I had the same thought and am +1 to this, `Accepted` seems a little bit easier to understand for users IMO and can be consistently used across all resources.

* In general, resources should be considered `Attached` if their config is valid
enough to generate some config in the underlying data plane.
* There will be a limited set of positive polarity summary conditions, and a number

> of other specific negative-polarity error conditions.

Reviewer 5: This will make UI display somewhat harder (green dots / checkmarks / exclamation points), but I realize it may align with human reading a bit more easily.

I don't have a strong opinion here, just want to highlight the cost elsewhere in the stack.

* All relevant positive-polarity summary Conditions for a resource must be added
when it's observed.

> For example, HTTPRoutes must always have `Accepted` and `ResolvedRefs`, regardless

Author: I hadn't updated, but with the change to `Accepted` everywhere, this is now correct, thanks.

of their state.
* Negative polarity error conditions must only be added when the error is True.
* The `Ready` condition will be moved to Extended conformance, and we'll reevaluate
if it's used by any implementations after some time has passed. If not, it may be
removed.
* To capture the behavior that `Ready` currently captures, `Programmed` will be
introduced. This means that the implementation has seen the config, has everything
it needs, parsed it, and sent configuration off to the data plane. The configuration
should be available "soon". We'll leave "soon" undefined for now.
* The summary conditions will be added into the CRD definitions

> as default values for `status.Conditions`, with a status of `Unknown`.

Author: I was intending to make it that the non-`controllerName` Conditions will be present, but I'll leave this out for now.

* Resolving a comment that came up, documentation will be added to clarify that

> it's okay to add your own Conditions, and that implementations should namespace

Reviewer 2: We should probably also recommend that implementations publish both new and old conditions for several months to provide a smoother transition.

can't be sure if everything is good unless you see `AllGood: true`. The absence
of a positive-polarity condition implies that the condition could be false. This
puts this option more in line with the API guidelines on this point.

#### Some Conditions Positive

In this case, only a limited set of summary conditions are positive, and the rest
are negative.

Pros:
* Error states can be described with `Error: true` instead of `NoError: false`.
* Negative polarity error conditions are more friendly to not being present (since
absence of `Error: true` implies everything's okay).

Cons:
* Any code handling conditions will need a list of the positive ones, and will
need to assume that any others are negative.

#### Decision

Gateway API conditions will be positive for conditions that describe the happy
state of the object, which is currently `Attached` and `ResolvedRefs`, and will 
also include the new `Programmed` condition, and the newly-Extended condition
`Ready`.

### Should conditions always be added?

Not all of them.

Positive polarity Conditions that describe the desirable state of the object must
always be set. These are currently `Attached`, `ResolvedRefs`, and `Programmed`.

> Implementations that use `Ready` must also add it immediately, set to `false`.

Author: Yeah, that seems like a good idea.

### New and Updated Conditions

#### `Attached`

This GEP proposes replacing all conditions that indicate syntactic and semantic
validity with one, `Attached` condition type, with the exception of the
GatewayClass resource (because it's the root of the resource tree and doesn't
attach to anything).

That is, the proposal is to replace:
* `Scheduled` on Gateway
* `Detached` on Listener
* `Accepted` on Route

with `Attached` in all these locations.

> GatewayClass will maintain the `Accepted` condition.

Author: Yeah, agreed. I've migrated the proposal to this.

All of these conditions share the following meanings:
* The resource has been accepted for processing by the controller

> * The resource is syntatically and semantically valid, and internally consistent

Reviewer 6: nit

```suggestion
* The resource is syntactically and semantically valid, and internally consistent
```

* The resource fits into a larger system of Gateway API resources, and there is
is no missing information, including but not limited to:
  * Any mandatory references resolve to existing resources (examples here are the
  Gateway's gatewayClass field, or the `parentRefs` field in Route resources)
  * Any specified TLS secrets exist
* The resource is supported by the controller by ensuring things like:
  * Any Kinds being referred to by the resource are supported
  * Features being used by the resource are supported

All of these rules can be summarized into:
* The resource is valid enough to produce some configuration in the underlying
data plane.

> For Gateway, `Attached` also subsumes the functions of `Scheduled`: `Attached`

Reviewer 4: On some impls, provisioning takes about a minute. That may mean that some errors that are immediately known are masked. Or I suppose they aren't masked, but we need to wait the entire minute to know if its valid. The message field could say "all fields valid, waiting for deployment" perhaps, but that isn't in the machine parse-able section.

That may not warrant change, just a thought

Author: I worded this carefully, so as not to imply that the provisioning needs to have _completed_, just that _it can be started_. The completion of infra provisioning should be handled by `Programmed` or `Ready`, or both.
